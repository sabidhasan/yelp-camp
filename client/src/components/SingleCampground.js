import React from 'react'
import CampMap from './CampMap'
import InfoBox from './InfoBox'
import SideBarInfoBox from './SideBarInfoBox'
import Reviews from './Reviews'
import RatingBar from './RatingBar'
import NewReviewForm from './NewReviewForm'

class SingleCampground extends React.Component {
  constructor(props) {
    super(props);
    this.requestedID = this.props.match.params.id;

    this.state = {
      editable: false,
      id: undefined,
      image: undefined,
      name: undefined,
      rating: 3.5,
      description: "Riverbend Campground is located on the Sheep River near Okotoks, Alberta. A small town just 20 kms from Calgary’s city limits. Riverbend is in the heart of Alberta’s ranchlands with a view of the spectacular Rocky Mountains and within a day trip to Banff National Park.",
      author: "Automatic",
      address: "48033 370 Ave E, Okotoks, AB  T1S 1B5",
      email: 'info@riverbendcampground.ca',
      lat: -34,
      lon: 150.66,
      phone: "(403) 938-2017",
      sites: 280,
      hours: {daily: '9 am to 9 pm (front gate closed at 11 pm)', seasonal: 'Open all year'},
      prices: {visitors: 2, daily: [40, 50], weekly: [270, 330], seasonal: null, description: 'Free for children 6 and under.'},
      paymentMethods: ['interac', 'cash'],
      reviews: [
        {id: 0, author: 'John', time: 'Apr 1, 2018', rating: 5, text: 'More text here'},
        {id: 1, author: 'Jane', time: 'Mar 28, 2018', rating: 4, text: 'Lorem Ipsum'},
        {id: 2, author: 'Jim', time: 'Mar 22, 2018', rating: 2, text: 'This is an OK campground'},
        {id: 3, author: 'Jason', time: 'Jan 2, 2018', rating: 0, text: 'Worst camp ever'}
      ]
    }

    this.editFormToggle = this.editFormToggle.bind(this);
  }

  componentDidMount(props) {
    fetch(`/campground?id=${this.requestedID}`)
      .then(res => res.json())
      .then(campground => {
        this.setState({id: campground.id, image: campground.image, name: campground.name});
      })
      .catch(err => {
        //TO--DO proper error handling for all fetch
        console.log("there is an error");
      });
  }

  editFormToggle(event) {
    event.preventDefault();
    const test = !this.state.editable;
    this.setState({editable: test}, () => console.log(this.state));
    // const newreviews = this.state.reviews.concat([{id: 4, author: 'test', time: '2', rating: '4', text: 'sometext'}])
    // this.setState({reviews: newreviews}, () => console.log(this.state, 'edit complete'));
    // this.setState({rating: 5})
  }

  render() {
    return (
      <div className='singleCampground'>
        <h1 className='title'>{this.state.name}</h1>

        <button href="#" className="review-link" onClick={this.editFormToggle} >Write a Review</button>

        <div className='rating'>
          <RatingBar rating={this.state.rating} />
          <h2>{this.state.reviews.length} Reviews</h2>
        </div>


        <CampMap
          googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyC4R6AN7SmujjPUIGKdyao2Kqitzr1kiRg&v=3.exp&libraries=geometry,drawing,places"
          loadingElement={<div className='map' />}
          containerElement={<div className='mapContainer' />}
          mapElement={<div className='map' />}
          lat={this.state.lat}
          lon={this.state.lon}
        />

        <InfoBox address={this.state.address} phone={this.state.phone} email={this.state.email} />

        <div className="description">
          <h1>Description</h1>
          {this.state.description}
        </div>

        <SideBarInfoBox hours={this.state.hours}
          campsites={this.state.sites}
          prices={this.state.prices}
          paymentMethods={this.state.paymentMethods}
        />

        <img className='campground-image' src={this.state.image} alt="" />

        <h1>Reviews</h1>
        <NewReviewForm editable={this.state.editable} editFormToggle={this.editFormToggle} />
        <Reviews reviews={this.state.reviews} />
      </div>
    )
  }
}

export default SingleCampground;
