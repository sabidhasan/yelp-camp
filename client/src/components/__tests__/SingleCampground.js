import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import SingleCampground from '../SingleCampground'

configure({ adapter: new Adapter() });

describe('<SingleCampground />', () => {
  let component;
  let props;
  let addToCart;
  let toggleLoginForm;
  const comments = [{rating: 4}, {rating: 2}, {rating: 1}, {rating: 4}];

  beforeEach(() => {
    toggleLoginForm = jest.fn();
    addToCart = jest.fn();
    props = {
      match: {
        params: {id: 10}
      },
      addToCart,
      toggleLoginForm
    }
    const state = {
      comments,
      name: 'Sample Name', description: 'This is a description of the campground',
      region: 'Region', province: 'Province'
    }
    component = shallow(
      <SingleCampground { ...props } />, {
        context: {
          startLoad: jest.fn(), finishLoad: jest.fn(), user: {loading: false}
        }
      });
    component.setState(state)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  })

  it('Clicking add to cart calls addToCart function', () => {
    component.find('.SingleCampground__cart').simulate('click')
    expect(addToCart).toHaveBeenCalledTimes(1);
  });

  it('SingleCampground__description exists when AJAX request is made', () => {
    expect(component.find('.SingleCampground__description')).toHaveLength(1);
  });

  it('Reviews are properly pluralized for >1 reviews', () => {
    expect(component.find('.SingleCampground__rating-count').text()).toBe('4 Reviews');
  });

  it('Reviews are properly pluralized for 1 review', () => {
    component.setState({
      comments: [{rating: 4}]
    });
    expect(component.find('.SingleCampground__rating-count').text()).toBe('1 Review');
  });

  it('Reviews are properly pluralized for 0 reviews', () => {
    component.setState({
      comments: []
    });
    expect(component.find('.SingleCampground__rating-count').text()).toBe('No Reviews');
  });

  it('Name is displayed', () => {
    expect(component.find('.SingleCampground__activities-header').text()).toBe('Activities at Sample Name');
  });

  it('Adding a new comment adds it to internal state', () => {
    component.instance().addNewComment({rating: 1})
    expect(component.state('comments').length).toBe(5);
  });

  it('RatingBar\'s rating props is calculated based on reviews', () => {
    const calculatedRating = comments.reduce((acc, val) => (acc + val.rating), 0) / comments.length
    expect(component.find('RatingBar').prop('rating')).toBe(calculatedRating)
  })

  it('NewReviewForm does not exist initially (form hidden on load)', () => {
    expect(component.find('NewReviewForm')).toHaveLength(0)
  })

  it('NewReviewForm exists when toggle button is clicked and user is logged in', () => {
    // Call toggle button press (done from child component in the app)
    component.instance().toggleReviewForm(null, true);
    expect(component.find('NewReviewForm')).toHaveLength(1)
  });

  it('NewReviewForm doesn\'t exist when toggle button is clicked and user is logged out', () => {
    // Rerender component with no user
    component = shallow(
      <SingleCampground { ...props } />, {
        context: {
          startLoad: jest.fn(), finishLoad: jest.fn(), user: undefined
        }
      });
    // Try to toggle review form
    component.instance().toggleReviewForm(null, true);
    // Review form should be hidden, since user isnt logged in
    expect(component.find('NewReviewForm')).toHaveLength(0);
  })

  it('toggleLoginForm function is called when toggle button is clicked and user is logged out', () => {
    // Rerender component with no user
    component = shallow(
      <SingleCampground { ...props } />, {
        context: {
          startLoad: jest.fn(), finishLoad: jest.fn(), user: undefined
        }
      });
    component.instance().toggleReviewForm(null, true);
    expect(toggleLoginForm).toHaveBeenCalledTimes(1);
  });

  it('SingleCampground component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
});
