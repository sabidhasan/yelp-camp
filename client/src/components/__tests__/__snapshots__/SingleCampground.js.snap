// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleCampground /> SingleCampground component snapshot matches what is expected 1`] = `
<div
  className="SingleCampground"
>
  <withStickyBar
    name="Sample Name"
    region="Region, Province"
  />
  <section
    className="SingleCampground__review-link"
  >
    <ReviewButton
      toggleReviewForm={[Function]}
    />
    <button
      aria-label="Add to Shopping Cart"
      className="SingleCampground__cart btn btn--flat"
      onClick={[Function]}
    >
      <i
        className="fas fa-shopping-cart"
      />
      Add to Cart
    </button>
  </section>
  <section
    className="SingleCampground__rating"
  >
    <RatingBar
      rating={2.75}
    />
    <h2
      className="SingleCampground__rating-count"
    >
      4
       Review
      s
    </h2>
  </section>
  <section
    className="CampMap"
  >
    <Wrapper />
  </section>
  <section
    className="SingleCampground__info-weather"
  >
    <InfoBox />
    <WeatherBox />
  </section>
  <section
    className="SingleCampground__description"
  >
    <h2>
      Description
    </h2>
    This is a description of the campground
  </section>
  <Sidebar />
  <LazyLoad
    offset={0}
    once={true}
    overflow={false}
    resize={false}
    scroll={true}
    unmountIfInvisible={false}
  >
    <img
      alt="Campground"
      className="SingleCampground__image"
    />
  </LazyLoad>
  <section
    className="SingleCampground__activities"
  >
    <h1
      className="SingleCampground__activities-header"
    >
      Activities at 
      Sample Name
    </h1>
    <Activities />
  </section>
  <section
    className="SingleCampground__reviews-header"
  >
    <h1>
      Reviews
    </h1>
    <ReviewButton
      toggleReviewForm={[Function]}
    />
  </section>
  <Reviews
    comments={
      Array [
        Object {
          "rating": 4,
        },
        Object {
          "rating": 2,
        },
        Object {
          "rating": 1,
        },
        Object {
          "rating": 4,
        },
      ]
    }
    deleteReview={[Function]}
  />
</div>
`;
